### PR TITLE
CMR-7185 autocomplete suggestion indexing logic to prevent unsearchable suggestions

### PR DESCRIPTION
**Pull request by jaybarra**

### CODE REVIEW

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 1]
```diff
@@ -292,77 +294,67 @@
   (try
     (cp/parse-concept context collection)
     (catch Exception e
-      (error (format "An error occurred while parsing collection: %s"
-                     (.getMessage e)))
-      nil)))
+      (error (format "An error occurred while parsing collection for autocomplete with concept-id [%s]: %s"
+                     (:concept-id collection)
+                     (.getMessage e))))))
```
Comment (mschmele) [inline indexer-app/src/cmr/indexer/services/index_service.clj:299]: 👏

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 2]
```diff
@@ -292,77 +294,67 @@
...
 
-(defn- collection->suggestion-doc
+(defn- collections->suggestion-docs
   "Convert collection concept metadata to UMM-C and pull facet fields
   to be indexed as autocomplete suggestion doc"
   [context collections provider-id]
-  (let [parse-coll (partial parse-collection context)
+  (let [{:keys [index-names]} (idx-set/get-concept-type-index-names context)
+        index (get-in index-names [:autocomplete :autocomplete])
```
Comment (mschmele) [inline indexer-app/src/cmr/indexer/services/index_service.clj:306]: This is definitely what I should have done from the beginning - nice

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 3]
```diff
@@ -292,77 +294,67 @@
...
+        parse-coll-fn (partial parse-collection context)
+        humanize-fields-fn (partial humanizer/collection-humanizers-elastic context)
         parsed-concepts (->> collections
                              (remove :deleted)
-                             (map parse-coll)
+                             (map parse-coll-fn)
```
Comment (mschmele) [inline indexer-app/src/cmr/indexer/services/index_service.clj:311]: I think I actually prefer just having this line be `(map (partial parse-collection context))`. Your eyes don't have to jump around the screen as much

**Review by mschmele — APPROVED** [review]

**Review by daniel-zamora — APPROVED** [review]

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 4]
```diff
@@ -189,14 +188,27 @@
 
 (def REINDEX_BATCH_SIZE 2000)
 
+(defn- remove-nil-tail
+  "Remove any nils from the end of a collection."
+  [coll]
+  (->> coll
```
Comment (ygliuvt) [inline indexer-app/src/cmr/indexer/services/index_service.clj:194]: reverse the collection twice to drop items from the end does not pass the smell test. Change it to something like:
```
(loop [x coll] 
  (if (last x) 
    x 
    (recur (drop-last x))))
```

**Review by mschmele — APPROVED** [review]

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 5]
```diff
@@ -189,14 +188,28 @@
 
 (def REINDEX_BATCH_SIZE 2000)
 
+(defn- remove-nil-tail
+  "Remove any nils from the end of a science-keyword hierarchy.
+  Nil values can exist in between the leaf keyword and Topic."
+  [coll]
+  (loop [x coll]
+    (if (last x)
```
Comment (ygliuvt) [inline indexer-app/src/cmr/indexer/services/index_service.clj:196]: change this line to `(if (or (last x) (empty? x))`

File: system-int-test/test/cmr/system_int_test/search/autocomplete/suggestion_reindex_test.clj [hunk 6]
```diff
@@ -177,91 +178,119 @@
...
+
            results (get-in (search/get-autocomplete-json (str "q=" query)) [:feed :entry])]
        (compare-autocomplete-results results expected))
      "None found"
      "stale" []
 
      "Still none found"
      "old" [])))
+
+(deftest semi-hierachical-keywords-test
```
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/search/autocomplete/suggestion_reindex_test.clj:272]: Add a test for collection with no science keywords

**Review by ygliuvt — APPROVED** [review]

**Pull request closed without merging**